core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -515,7 +515,11 @@ class ReplicaManager(val config: KafkaConfig,
             case Some(p) =>
               if (p eq ReplicaManager.OfflinePartition)
                 throw new KafkaStorageException("Partition %s is in an offline log directory on broker %d".format(topicPartition, localBrokerId))
-              p
+              p.leaderReplicaIfLocal match {
+                case Some(_) => p
+                case None =>
+                  throw new NotLeaderForPartitionException(s"Broker $localBrokerId is not leader for partition $topicPartition")
+              }
```

Comment:
We can use `getLeaderReplicaIfLocal` to simplify the code.

Reply:
getLeaderReplicaIfLocal returns replica.
The return value here is partition.
I feel the current formation is clearer.
See the code for convertedOffset below which has the same structure.

Reply:
The problem is that the same logic exists in many places leading to bugs. The `convertedOffset` code only needs a replica. However, it's true that we need the `partition` for `deleteRecordsOnLeader`. My suggestion is to introduce a new method and use it in `getReplicaLeaderIfLocal`:

```scala
def getLeaderReplicaIfLocal(topicPartition: TopicPartition): Replica =  {
    val (_, replica) = getPartitionAndLeaderReplicaIfLocal(topicPartition)
    replica
  }

  def getPartitionAndLeaderReplicaIfLocal(topicPartition: TopicPartition): (Partition, Replica) =  {
    val partitionOpt = getPartition(topicPartition)
    partitionOpt match {
      case None =>
        throw new UnknownTopicOrPartitionException(s"Partition $topicPartition doesn't exist on $localBrokerId")
      case Some(partition) =>
        if (partition eq ReplicaManager.OfflinePartition)
          throw new KafkaStorageException(s"Partition $topicPartition is in an offline log directory on broker $localBrokerId")
        else partition.leaderReplicaIfLocal match {
          case Some(leaderReplica) => (partition, leaderReplica)
          case None =>
            throw new NotLeaderForPartitionException(s"Leader not local for partition $topicPartition on broker $localBrokerId")
        }
    }
  }
```